File: tools/check_parens.js
const fs = require('fs');
const s = fs.readFileSync('static/app.js','utf8');
const stack = [];
const pairs = {')':'(',']':'[','}':'{'};
for(let i=0;i<s.length;i++){
  const ch = s[i];
  if('([{'.includes(ch)) stack.push({ch,i:i+1});
  else if(')]}'.includes(ch)){
    if(stack.length===0){ console.log('Unmatched',ch,'at',i+1); process.exit(0); }
    const top = stack.pop();
    if(top.ch !== pairs[ch]){ console.log('Mismatched',top.ch,'vs',ch,'at',i+1); process.exit(0); }
  }
}
if(stack.length) console.log('Unclosed at end, top=',stack[stack.length-1]); else console.log('All balanced');

File: tools/check_parens.py
from pathlib import Path
s = Path('static/app.js').read_text(encoding='utf-8')
stack = []
pairs = {')':'(',']':'[','}':'{'}
for i, ch in enumerate(s):
    if ch in '([{':
        stack.append((ch, i+1))
    elif ch in ')]}':
        if not stack:
            print('Unmatched', ch, 'at', i+1)
            break
        top, idx = stack.pop()
        if top != pairs[ch]:
            print('Mismatched', top, 'vs', ch, 'at', i+1)
            break
else:
    if stack:
        print('Unclosed at end, top=', stack[-1])
    else:
        print('All balanced')
